### PR TITLE
Correct Forward pass and gradient test passing

### PR DESCRIPTION
**Pull request by tomiock**
- decorated function used for the forward pass
- correct model initialization
- explicit computation graph construction according to first forward pass
- gradient test passing

### CODE REVIEW

File: macrograd/engine.py [hunk 1]
```diff
@@ -527,6 +537,17 @@ def visualize(
...
+            if node.type == NodeType.DATA:
+                label = f"ID: {node.id!r}\nop: {op_name}\nshape: {node.shape}\n{node.type}\ndtype: {node.dtype}\nrg:{node.requires_grad}"
+
+                color_node = "red"
+            if node.type == NodeType.PARAM:
+                color_node = "blue"
+            if node.type == NodeType.CONST:
+                color_node = "green"
+            if node.type == NodeType.COMPUTED:
+                color_node = "grey"
```
Comment (tomiock) [inline macrograd/engine.py:549]: dogshit code here
